server/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -696,6 +696,46 @@ static String getCommitCheckpointFileName(long generation) {
         return TRANSLOG_FILE_PREFIX + generation + CHECKPOINT_SUFFIX;
     }
 
+    /**
+     * Trims translog (effectively moves max visible seq# {@link Checkpoint#trimmedAboveSeqNo}) for terms below <code>belowTerm</code>
+     * and seq# above <code>aboveSeqNo</code>
+     */
+    public void trim(long belowTerm, long aboveSeqNo) throws IOException {
+        if (aboveSeqNo < 0) {
+            return;
+        }
+
+        final ChannelFactory channelFactory = FileChannel::open;
+        try (ReleasableLock lock = writeLock.acquire()) {
+
+            // update all existed ones (if it is necessary) as checkpoint and reader are immutable
+            final List<TranslogReader> newReaders = new ArrayList<>(readers.size());
+            for (TranslogReader reader : readers) {
+                final Checkpoint checkpoint = reader.getCheckpoint();
+                final Path checkpointFile = location.resolve(getCommitCheckpointFileName(checkpoint.generation));
+
+                final TranslogReader newReader;
+                if (reader.getPrimaryTerm() < belowTerm
+                    && (aboveSeqNo < checkpoint.trimmedAboveSeqNo || checkpoint.trimmedAboveSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO)) {
+                    final Checkpoint newCheckpoint = new Checkpoint(checkpoint.offset, checkpoint.numOps,
+                        checkpoint.generation, checkpoint.minSeqNo, checkpoint.maxSeqNo,
+                        checkpoint.globalCheckpoint, checkpoint.minTranslogGeneration, aboveSeqNo);
+                    Checkpoint.write(channelFactory, checkpointFile, newCheckpoint, StandardOpenOption.WRITE);
+
+                    IOUtils.fsync(checkpointFile, false);
+                    IOUtils.fsync(checkpointFile.getParent(), true);
+
+                    newReader = reader.withNewCheckpoint(newCheckpoint);
```

Comment:
this is super dangerous. If we have an exception here (or in pervious iterations) we will leak a file channel, as no one closes `newReader`. I think you need testing for this, see `TranslogTests#testWithRandomException` to see how this can be tested.

Reply:
agreed, going to fix it